File: app/src/main/java/org/illegaller/ratabb/hishoot2i/view/fragment/templateview/TemplateAdapter.java
package org.illegaller.ratabb.hishoot2i.view.fragment.templateview;

import android.os.Bundle;
import android.support.v7.widget.RecyclerView;
import android.view.ViewGroup;
import com.chauthai.swipereveallayout.ViewBinderHelper;
import java.util.ArrayList;
import java.util.List;
import javax.inject.Inject;
import org.illegaller.ratabb.hishoot2i.model.template.Template;

public class TemplateAdapter extends RecyclerView.Adapter<TemplateViewHolder> {
  private final List<Template> mTemplateList = new ArrayList<>();
  private final List<String> mFavList = new ArrayList<>();
  private final ViewBinderHelper mBinderHelper = new ViewBinderHelper();

  @Inject TemplateAdapter() {
  }

  void setFavList(List<String> mFavList) {
    this.mFavList.clear();
    this.mFavList.addAll(mFavList);
  }

  @Override public TemplateViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
    return TemplateViewHolder.inflate(parent);
  }

  @Override public void onBindViewHolder(TemplateViewHolder holder, int position) {
    holder.onBind(new TemplateAndFavList(getItem(position), mFavList));
    mBinderHelper.bind(holder.swipeRevealLayout, getItem(position).id);
  }

  @Override public int getItemCount() {
    return mTemplateList.size();
  }

  void saveStates(Bundle bundle) {
    mBinderHelper.saveStates(bundle);
  }

  void restoreStates(Bundle bundle) {
    mBinderHelper.restoreStates(bundle);
  }

  void animateTo(List<Template> templates) {
    applyAndAnimateRemovals(templates);
    applyAndAnimateAdditions(templates);
    applyAndAnimateMovedItems(templates);
  }

  private Template getItem(int position) {
    return mTemplateList.get(position);
  }

  private Template removeItem(int position) {
    final Template template = mTemplateList.remove(position);
    notifyItemRemoved(position);
    return template;
  }

  private void applyAndAnimateRemovals(List<Template> newList) {
    for (int i = mTemplateList.size() - 1; i >= 0; i--) {
      final Template template = mTemplateList.get(i);
      if (!newList.contains(template)) removeItem(i);
    }
  }

  private void applyAndAnimateAdditions(List<Template> newList) {
    for (int i = 0; i < newList.size(); i++) {
      final Template template = newList.get(i);
      if (!mTemplateList.contains(template)) addItem(i, template);
    }
  }

  private void applyAndAnimateMovedItems(List<Template> newList) {
    for (int to = newList.size() - 1; to >= 0; to--) {
      final Template model = newList.get(to);
      final int from = mTemplateList.indexOf(model);
      if (from >= 0 && from != to) moveItem(from, to);
    }
  }

  private void addItem(int position, Template template) {
    mTemplateList.add(position, template);
    notifyItemInserted(position);
  }

  private void moveItem(int from, int to) {
    final Template template = mTemplateList.remove(from);
    mTemplateList.add(to, template);
    notifyItemMoved(from, to);
  }
}

File: app/src/main/java/org/illegaller/ratabb/hishoot2i/ui/template/fragment/TemplateAdapter.kt
package org.illegaller.ratabb.hishoot2i.ui.template.fragment

import android.support.v7.widget.RecyclerView
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import org.illegaller.ratabb.hishoot2i.R
import org.illegaller.ratabb.hishoot2i.ui.common.asyncListDiffer
import rbb.hishoot2i.common.ext.graphics.color
import rbb.hishoot2i.common.ext.inflateNotAttach
import rbb.hishoot2i.common.ext.isVisible
import rbb.hishoot2i.common.ext.preventMultipleClick
import rbb.hishoot2i.common.ext.toDateTimeFormat
import rbb.hishoot2i.common.imageloader.ImageLoader
import rbb.hishoot2i.template.Template
import javax.inject.Inject

class TemplateAdapter @Inject constructor(
    private val imageLoader: ImageLoader,
    templateDiffCallback: TemplateDiffCallback
) : RecyclerView.Adapter<TemplateAdapter.TemplateHolder>() {

    init {
        setHasStableIds(true)
    }

    private val differ = asyncListDiffer(templateDiffCallback)

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): TemplateHolder =
        with(parent) { inflateNotAttach(R.layout.row_item_template).let { TemplateHolder(it) } }

    override fun getItemId(position: Int): Long =
        differ.currentList[position].id.hashCode().toLong()

    override fun getItemCount(): Int = differ.currentList.size
    override fun onBindViewHolder(holder: TemplateHolder, position: Int) {
        holder.bind(differ.currentList[position])
    }

    internal var clickItem: (Template) -> Unit = { _: Template -> }
    @Throws(IndexOutOfBoundsException::class)
    fun getItemAt(position: Int): Template = differ.currentList[position]

    @Throws(IndexOutOfBoundsException::class)
    fun removeItemAt(position: Int): Template {
        var itemRemoved = getItemAt(position)
        val list = differ.currentList.toMutableList().apply {
            itemRemoved = removeAt(position)
        }
        submitList(list, true)
        return itemRemoved
    }

    fun submitList(newList: List<Template>, isForceUpdateUI: Boolean) {
        differ.submitList(newList)
        if (isForceUpdateUI) notifyItemRangeChanged(0, itemCount) //
    }

    inner class TemplateHolder(view: View) : RecyclerView.ViewHolder(view) {
        private val colorBackFav: Int = itemView.context.color(R.color.accent)
        private val colorBackDelete: Int = itemView.context.color(R.color.red)
        private val itemTemplateBack: View = itemView.findViewById(R.id.itemTemplateBack)
        private val itemIndicatorFav: View = itemView.findViewById(R.id.itemIndicatorFav)
        private val itemIndicatorDelete: View = itemView.findViewById(R.id.itemIndicatorDelete)
        private val itemTemplatePreview: ImageView = itemView.findViewById(R.id.itemTemplatePreview)
        private val itemTemplateName: TextView = itemView.findViewById(R.id.itemTemplateName)
        private val itemTemplateId: TextView = itemView.findViewById(R.id.itemTemplateId)
        private val itemTemplateInfo: TextView = itemView.findViewById(R.id.itemTemplateInfo)
        /*SwipeHelper*/
        internal val foregroundView: View = itemView.findViewById(R.id.itemTemplateForeground)

        init {
            itemView.setOnClickListener {
                it.preventMultipleClick {
                    clickItem(differ.currentList[adapterPosition])
                }
            }
        }

        /* SwipeHelper: Template Default can not deleted. */
        fun isItemDeletable(): Boolean = !differ.currentList[adapterPosition].isDefault

        fun handleIndicatorVisible(dX: Float, isActive: Boolean) {
            itemTemplateBack.isVisible = isActive
            val isFav = dX > 0
            itemIndicatorFav.isVisible = isFav
            itemIndicatorDelete.isVisible = !isFav
            (if (isFav) colorBackFav else colorBackDelete).let {
                itemTemplateBack.setBackgroundColor(it)
            }
        }

        fun bind(item: Template) {
            imageLoader.display(itemTemplatePreview, item.preview)
            itemTemplateName.text = item.name
            itemTemplateId.text = item.id
            with(itemTemplateInfo) {
                text = context.getString(
                    R.string.template_info_format,
                    item.author,
                    item.desc,
                    item.installedDate.toDateTimeFormat()
                )
            }
        }
    }
}